File: apps/web/scripts/agent_harness/sync-skills.mjs
#!/usr/bin/env bun

import { cp, mkdir, readdir, readFile, rm } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const SCRIPT_DIR = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(SCRIPT_DIR, "../..");
const DEFAULT_SOURCE = path.join(ROOT, ".agents/skills");
const DEFAULT_MIRROR = path.join(ROOT, ".claude/skills");

async function treeFiles(root) {
	const files = new Map();

	async function visit(directory) {
		for (const entry of await readdir(directory, { withFileTypes: true })) {
			const absolute = path.join(directory, entry.name);
			if (entry.isDirectory()) {
				await visit(absolute);
			} else if (entry.isFile()) {
				files.set(path.relative(root, absolute).split(path.sep).join("/"), await readFile(absolute));
			}
		}
	}

	try {
		await visit(root);
	} catch (error) {
		if (error?.code !== "ENOENT") {
			throw error;
		}
	}
	return files;
}

export async function compareTrees(source = DEFAULT_SOURCE, mirror = DEFAULT_MIRROR) {
	const [sourceFiles, mirrorFiles] = await Promise.all([treeFiles(source), treeFiles(mirror)]);
	const errors = [];

	for (const [relative, content] of sourceFiles) {
		if (!mirrorFiles.has(relative)) {
			errors.push(`missing from Claude mirror: ${relative}`);
		} else if (!content.equals(mirrorFiles.get(relative))) {
			errors.push(`content drift: ${relative}`);
		}
	}
	for (const relative of mirrorFiles.keys()) {
		if (!sourceFiles.has(relative)) {
			errors.push(`extra in Claude mirror: ${relative}`);
		}
	}
	return errors.sort();
}

export async function syncTrees(source = DEFAULT_SOURCE, mirror = DEFAULT_MIRROR) {
	await rm(mirror, { recursive: true, force: true });
	await mkdir(path.dirname(mirror), { recursive: true });
	await cp(source, mirror, { recursive: true });
}

async function main() {
	if (process.argv.includes("--check")) {
		const errors = await compareTrees();
		if (errors.length > 0) {
			process.stderr.write(`${errors.join("\n")}\n`);
			process.exitCode = 1;
			return;
		}
		process.stdout.write("Agent skill mirror is synchronized.\n");
		return;
	}

	await syncTrees();
	process.stdout.write("Synchronized .agents/skills to .claude/skills.\n");
}

if (import.meta.main) {
	await main();
}

File: tooling/agent_harness/sync_skills.py
#!/usr/bin/env python3
"""Mirror canonical project skills into Claude Code's discovery directory."""

from __future__ import annotations

import argparse
import os
import shutil
from dataclasses import dataclass
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
SOURCE = ROOT / ".agents" / "skills"
MIRROR = ROOT / ".claude" / "skills"


@dataclass(frozen=True)
class TreeEntry:
    """Portable file identity required for a trustworthy generated mirror."""

    kind: str
    content: bytes
    executable: bool


def tree_files(root: Path) -> dict[str, TreeEntry]:
    """Return deterministic content, kind, and executable-bit identities."""
    if not root.is_dir():
        return {}
    entries: dict[str, TreeEntry] = {}
    for path in sorted(root.rglob("*")):
        if path.is_symlink():
            entry = TreeEntry(
                kind="symlink",
                content=os.fsencode(path.readlink()),
                executable=False,
            )
        elif path.is_file():
            entry = TreeEntry(
                kind="file",
                content=path.read_bytes(),
                executable=bool(path.stat().st_mode & 0o111),
            )
        else:
            continue
        entries[path.relative_to(root).as_posix()] = entry
    return entries


def compare_trees(source: Path = SOURCE, mirror: Path = MIRROR) -> list[str]:
    """Describe missing, extra, and changed files between two skill trees."""
    source_files = tree_files(source)
    mirror_files = tree_files(mirror)
    problems: list[str] = []
    for relative in sorted(source_files.keys() - mirror_files.keys()):
        problems.append(f"missing from Claude mirror: {relative}")
    for relative in sorted(mirror_files.keys() - source_files.keys()):
        problems.append(f"extra in Claude mirror: {relative}")
    for relative in sorted(source_files.keys() & mirror_files.keys()):
        canonical = source_files[relative]
        generated = mirror_files[relative]
        if canonical.kind != generated.kind:
            problems.append(f"kind drift: {relative}")
        elif canonical.executable != generated.executable:
            problems.append(f"executable mode drift: {relative}")
        elif canonical.content != generated.content:
            problems.append(f"content drift: {relative}")
    return problems


def sync(source: Path = SOURCE, mirror: Path = MIRROR) -> None:
    """Replace the generated mirror after validating the canonical source."""
    if not source.is_dir() or not any(source.iterdir()):
        raise SystemExit(f"canonical skill directory is missing or empty: {source}")
    if mirror.exists():
        shutil.rmtree(mirror)
    shutil.copytree(source, mirror, symlinks=True, copy_function=shutil.copy2)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument(
        "--check", action="store_true", help="fail if the mirror differs"
    )
    args = parser.parse_args()

    if args.check:
        problems = compare_trees()
        if problems:
            print("Agent skill mirror is out of sync:")
            for problem in problems:
                print(f"- {problem}")
            return 1
        print("Agent skill mirror is synchronized.")
        return 0

    sync()
    print(f"Mirrored {SOURCE.relative_to(ROOT)} -> {MIRROR.relative_to(ROOT)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
